### PR TITLE
fix #1009: New instance of home created only if language is changed

### LINKED ISSUE #1009: Back from Settings Activity creates new HomeActivity
**Summary:**

When we go to Settings menu and click back the app creates a new activity. Exiting from this instance of HomeActivity will take you to previous one
The issue was raised before  #891. But the bug is still there.

**Steps to reproduce:**

Go to Settings Menu. Go back to return to HomeActivity. Now press back 2 more times. You will see that the current HomeActitviy closes and a new one opens.

**Expected behaviour:**

It should return to the already created HomeActivity.

**Observed behaviour:**

A new HomeActivity is created on clicking on back button.

**Device and Android version:**

Moto G5 Plus, Android 8.1

**Screenshots:**

ezgif-4-fe997b4891c7

### PR DESCRIPTION
**Pull request by rahul-jha98**
Fixes #1009

Please Add Screenshots If there are any UI changes.

- [x] Apply the `AndroidStyle.xml` style template to your code in Android Studio.

- [x] Run the unit tests with `./gradlew check` to make sure you didn't break anything

- [x] If you have multiple commits please combine them into one commit by squashing them.

### CODE REVIEW

File: app/src/main/java/org/mifos/mobilebanking/ui/activities/SettingsActivity.java [hunk 1]
```diff
@@ -16,12 +18,24 @@ protected void onCreate(Bundle savedInstanceState) {
...
+        }
     }
 
     @Override
     public void onBackPressed() {
         super.onBackPressed();
-        Intent i = new Intent(this, HomeActivity.class);
-        startActivity(i);
+        if (hasSettingsChanged) {
+            // If Settings have changed, the previous home activity may not be valid
```
Comment (miPlodder) [inline app/src/main/java/org/mifos/mobilebanking/ui/activities/SettingsActivity.java]: Remove the comments.

File: app/src/main/java/org/mifos/mobilebanking/ui/fragments/SettingsFragment.java [hunk 2]
```diff
@@ -68,7 +68,9 @@ public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, Strin
         if (preference instanceof ListPreference) {
             ListPreference listPreference = (ListPreference) preference;
             LanguageHelper.setLocale(getContext(), listPreference.getValue());
-            startActivity(new Intent(getActivity(), getActivity().getClass()));
+            Intent intent = new Intent(getActivity(), getActivity().getClass());
+            intent.putExtra("hasSettingsChanged", true);
```
Comment (miPlodder) [inline app/src/main/java/org/mifos/mobilebanking/ui/fragments/SettingsFragment.java]: Don't use hard-code strings.

File: app/src/main/java/org/mifos/mobilebanking/ui/activities/SettingsActivity.java [hunk 3]
```diff
@@ -16,12 +18,24 @@ protected void onCreate(Bundle savedInstanceState) {
...
+            // For example if user changes the language we need to
+            // relaunch previous activity again
+
+            ActivityCompat.finishAffinity(this);
```
Comment (miPlodder) [inline app/src/main/java/org/mifos/mobilebanking/ui/activities/SettingsActivity.java:33]: I guess this piece of code is not needed as on pressing the Back button the Activity ends.

File: app/src/main/java/org/mifos/mobilebanking/ui/activities/SettingsActivity.java [hunk 4]
```diff
@@ -16,12 +18,24 @@ protected void onCreate(Bundle savedInstanceState) {
         setToolbarTitle(getString(R.string.settings));
         showBackButton();
         replaceFragment(SettingsFragment.newInstance(), false, R.id.container);
+        Intent intent = getIntent();
+        if (intent.hasExtra("hasSettingsChanged")) {
```
Comment (miPlodder) [inline app/src/main/java/org/mifos/mobilebanking/ui/activities/SettingsActivity.java]: Don't use hard-coded strings. Instead, define constants in `Constants.java` and Don't create an object for Intent. Just use `getIntent()` and make the method calls.

File: app/src/main/java/org/mifos/mobilebanking/ui/activities/SettingsActivity.java [hunk 5]
```diff
@@ -2,12 +2,14 @@
...
 import android.content.Intent;
 import android.os.Bundle;
+import android.support.v4.app.ActivityCompat;
 
 import org.mifos.mobilebanking.R;
 import org.mifos.mobilebanking.ui.activities.base.BaseActivity;
 import org.mifos.mobilebanking.ui.fragments.SettingsFragment;
 
 public class SettingsActivity extends BaseActivity {
+    private boolean hasSettingsChanged;
```
Comment (miPlodder) [inline app/src/main/java/org/mifos/mobilebanking/ui/activities/SettingsActivity.java:14]: Add an empty line before this line.

**Review by miPlodder — CHANGES_REQUESTED** [review]

File: app/src/main/java/org/mifos/mobilebanking/ui/activities/SettingsActivity.java (hunk 3 above)
Comment (rahul-jha98) [reply app/src/main/java/org/mifos/mobilebanking/ui/activities/SettingsActivity.java:33]: Yes, but I need to call finishAffinity() for it to work the way I expected. 

Basically, what I have done is that if the user changes the preference for language the hasSettingsChanged boolean becomes true.
If it is true, then after the onbackPressed() call I need to  finish all activities i.e. Settings as well as Home Activity and then launch a new instance of HomeActivity. This new instance will be in the user selected language.

**Comment by miPlodder** [comment]
@rahul-jha98 Add GIF for this (every UI changes). Use DU recorder in your device for creating one. Rest looks good. 👍

**Comment by rahul-jha98** [comment]
When the user **changes the language settings** a new instance of Home Activity is created =>
changed

When the user **simply presses back with no change in language** the Settings Activity simply closes with no wasteful Activity creation.
nochange

**Comment by miPlodder** [comment]
@rahul-jha98  Good job 👍

**Pull request merged**